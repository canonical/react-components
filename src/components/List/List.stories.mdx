import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import List from "./List";

<Meta
  title="List"
  component={List}
  argTypes={{
    items: {
      control: {
        disable: true,
      },
    },
    stepped: {
      control: {
        disable: true,
      },
    },
  }}
/>

export const Template = (args) => <List {...args} />;

### List

This is a [React](https://reactjs.org/) component for the Vanilla [List](https://docs.vanillaframework.io/patterns/lists/).

If you want to display lists in a way that is more visually distinctive than the standard `<ol>` and `<ul>`, we have 7 list styles at your disposal.

### Props

<ArgsTable of={List} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{ items: ["Digital signage", "Robotics", "Gateways"] }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Ticked

<Canvas>
  <Story name="Ticked">
    <List
      items={[
        "Hardware guidance and sizing",
        "Fixed-price deployment",
        "Reference architecture",
      ]}
      ticked
    />
  </Story>
</Canvas>

### Horizontal divider

<Canvas>
  <Story name="Horizontal divider">
    <List
      items={[
        "Single-command install and upgrade",
        "Postgres database for reliable state",
        "High availability for all services",
      ]}
      divided
    />
  </Story>
</Canvas>

### Ticked with horizontal divider

<Canvas>
  <Story name="Ticked with horizontal divider">
    <List
      items={[
        "Balance compute load in the cloud",
        "Optimise for specific workloads",
        "Assess overcommit ratios",
      ]}
      divided
      ticked
    />
  </Story>
</Canvas>

### Inline

<Canvas>
  <Story name="Inline">
    <List items={["Community", "Careers", "Press centre"]} inline />
  </Story>
</Canvas>

### Middot

<Canvas>
  <Story name="Middot">
    <List
      items={["Legal information", "Data privacy", "Report a bug on this site"]}
      middot
    />
  </Story>
</Canvas>

### Stretch

<Canvas>
  <Story name="Stretch">
    <List
      items={[
        { content: <h3>Documentation</h3> },
        {
          content: <span className="p-label--positive">New</span>,
          className: "u-vertically-center u-align--right",
        },
      ]}
      stretch
    />
  </Story>
</Canvas>

### Vertical stepped

<Canvas>
  <Story name="Vertical stepped">
    <List
      items={[
        {
          title: "Log in to JAAS",
          content:
            "Ensure you have an Ubuntu SSO account before contacting JAAS. Log in to JAAS now.",
        },
        {
          title: "Configure a model",
          content:
            'Applications are contained within models and are installed via charms. Configure your model by pressing the "Start a new model" button.',
        },
        {
          title: "Credentials and SSH keys",
          content:
            "After having selected a cloud, a form will appear for submitting your credentials to JAAS. The below resources are available if you need help with gathering credentials.",
        },
      ]}
      stepped
    />
    <List
      items={[
        {
          title: "Log in to JAAS",
          content:
            "Ensure you have an Ubuntu SSO account before contacting JAAS. Log in to JAAS now.",
          titleElement: "h4",
        },
        {
          title: "Configure a model",
          content:
            'Applications are contained within models and are installed via charms. Configure your model by pressing the "Start a new model" button.',
          titleElement: "h4",
        },
        {
          title: "Credentials and SSH keys",
          content:
            "After having selected a cloud, a form will appear for submitting your credentials to JAAS. The below resources are available if you need help with gathering credentials.",
          titleElement: "h4",
        },
      ]}
      stepped
    />
  </Story>
</Canvas>

### Horizontal stepped

<Canvas>
  <Story name="Horizontal stepped">
    <List
      items={[
        {
          title: "Log in to JAAS",
          content:
            "Ensure you have an Ubuntu SSO account before contacting JAAS. Log in to JAAS now.",
        },
        {
          title: "Configure a model",
          content:
            'Applications are contained within models and are installed via charms. Configure your model by pressing the "Start a new model" button.',
        },
      ]}
      stepped
      detailed
    />
  </Story>
</Canvas>

### Split

<Canvas>
  <Story name="Split">
    <List
      items={[
        "Jointly shape the OpenStack architecture",
        "We help you plan your cloud hardware requirements",
        "We build OpenStack in your data center",
        "We operate the cloud to an SLA",
        "Transparent audit, logging, monitoring and management",
        "When your team is ready, we hand over the keys",
      ]}
      divided
      split
      ticked
    />
  </Story>
</Canvas>
