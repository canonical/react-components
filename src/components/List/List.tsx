import classNames from "classnames";
import React from "react";
import type { HTMLProps, ReactNode } from "react";

import { ClassName, Headings, PropsWithSpread } from "types";

export type ListItem =
  | ReactNode
  | PropsWithSpread<
      {
        content: ReactNode;
      },
      HTMLProps<HTMLLIElement>
    >;

export type SteppedListItem = PropsWithSpread<
  {
    content: ReactNode;
    title: ReactNode;
    titleElement?: Headings;
  },
  HTMLProps<HTMLLIElement>
>;

export type Props = {
  /**
   * Optional class(es) to pass to the wrapping element.
   */
  className?: ClassName;
  /**
   * The list's items.
   */
  items: ListItem[] | SteppedListItem[];
  detailed?: boolean;
  divided?: boolean;
  inline?: boolean;
  isDark?: boolean;
  middot?: boolean;
  split?: boolean;
  stepped?: boolean;
  stretch?: boolean;
  ticked?: boolean;
} & Omit<HTMLProps<HTMLOListElement>, "type">;

const generateItems = ({
  items,
  ticked,
  inline,
  middot,
  stepped,
}: Pick<Props, "items" | "ticked" | "inline" | "middot" | "stepped">) =>
  items.map((item, i) => {
    let body: ReactNode;
    let title: ReactNode;
    let className: string;
    let content: ReactNode;
    let TitleComponent: Headings = "h3";
    let liProps: HTMLProps<HTMLLIElement>;
    if (
      React.isValidElement(item) ||
      typeof item === "string" ||
      typeof item === "number"
    ) {
      content = item;
    } else if (item && typeof item === "object" && "content" in item) {
      if ("titleElement" in item) {
        ({
          title,
          className,
          content,
          titleElement: TitleComponent = "h3",
          ...liProps
        } = item);
      } else {
        ({ title, className, content, ...liProps } = item);
      }
    }
    if (stepped) {
      body = (
        <>
          <TitleComponent className="p-stepped-list__title">
            {title}
          </TitleComponent>
          <div className="p-stepped-list__content">{content}</div>
        </>
      );
    } else {
      body = content;
    }
    return (
      <li
        className={classNames(className, {
          "p-list__item": !inline && !stepped,
          "p-inline-list__item": inline,
          "p-stepped-list__item": stepped,
          "is-ticked": ticked,
        })}
        key={i}
        {...liProps}
      >
        {body}
        {/* middot variant in Vanilla relies on whitespace between inline elements for correct spacing */}
        {middot ? " " : null}
      </li>
    );
  });

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [List](https://docs.vanillaframework.io/patterns/lists/).
 *
 * If you want to display lists in a way that is more visually distinctive than the standard `<ol>` and `<ul>`, we have 7 list styles at your disposal.
 */
const List = ({
  className,
  detailed,
  divided,
  inline,
  isDark,
  items,
  middot,
  stretch,
  split,
  stepped,
  ticked,
  ...props
}: Props): JSX.Element => {
  const Component = stepped ? "ol" : "ul";
  return (
    <Component
      className={classNames(className, {
        "p-list": !divided && !inline && !middot && !stretch && !stepped,
        "p-list--divided": divided,
        "p-inline-list": inline,
        "p-inline-list--middot": middot,
        "p-inline-list--stretch": stretch,
        "p-stepped-list": stepped && !detailed,
        "p-stepped-list--detailed": stepped && detailed,
        "is-dark": isDark,
        "is-split": split,
      })}
      {...props}
    >
      {generateItems({
        items,
        ticked,
        inline: inline || middot || stretch,
        middot,
        stepped,
      })}
    </Component>
  );
};

export default List;
