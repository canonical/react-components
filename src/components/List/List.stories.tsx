import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import List from "./List";
import StatusLabel from "../StatusLabel";

const meta: Meta<typeof List> = {
  component: List,
  tags: ["autodocs"],

  argTypes: {
    items: {
      control: {
        disable: true,
      },
    },

    stepped: {
      control: {
        disable: true,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof List>;

export const Default: Story = {
  name: "Default",

  args: {
    items: ["Digital signage", "Robotics", "Gateways"],
  },
};

export const Ticked: Story = {
  render: () => (
    <List
      items={[
        "Hardware guidance and sizing",
        "Fixed-price deployment",
        "Reference architecture",
      ]}
      ticked
    />
  ),

  name: "Ticked",
};

export const HorizontalDivider: Story = {
  render: () => (
    <List
      items={[
        "Single-command install and upgrade",
        "Postgres database for reliable state",
        "High availability for all services",
      ]}
      divided
    />
  ),

  name: "Horizontal divider",
};

export const TickedWithHorizontalDivider: Story = {
  render: () => (
    <List
      items={[
        "Balance compute load in the cloud",
        "Optimise for specific workloads",
        "Assess overcommit ratios",
      ]}
      divided
      ticked
    />
  ),

  name: "Ticked with horizontal divider",
};

export const Inline: Story = {
  render: () => (
    <List items={["Community", "Careers", "Press centre"]} inline />
  ),
  name: "Inline",
};

export const Middot: Story = {
  render: () => (
    <List
      items={["Legal information", "Data privacy", "Report a bug on this site"]}
      middot
    />
  ),

  name: "Middot",
};

export const Stretch: Story = {
  render: () => (
    <List
      items={[
        {
          content: <h3>Documentation</h3>,
        },
        {
          content: <StatusLabel appearance="positive">New</StatusLabel>,
          className: "u-vertically-center u-align--right",
        },
      ]}
      stretch
    />
  ),

  name: "Stretch",
};

export const VerticalStepped: Story = {
  render: () => (
    <List
      items={[
        {
          title: "Log in to JAAS",
          content:
            "Ensure you have an Ubuntu SSO account before contacting JAAS. Log in to JAAS now.",
        },
        {
          title: "Configure a model",
          content:
            'Applications are contained within models and are installed via charms. Configure your model by pressing the "Start a new model" button.',
        },
        {
          title: "Credentials and SSH keys",
          content:
            "After having selected a cloud, a form will appear for submitting your credentials to JAAS. The below resources are available if you need help with gathering credentials.",
        },
      ]}
      stepped
    />
  ),

  name: "Vertical stepped",
};

export const HorizontalStepped: Story = {
  render: () => (
    <List
      items={[
        {
          title: "Log in to JAAS",
          content:
            "Ensure you have an Ubuntu SSO account before contacting JAAS. Log in to JAAS now.",
        },
        {
          title: "Configure a model",
          content:
            'Applications are contained within models and are installed via charms. Configure your model by pressing the "Start a new model" button.',
        },
      ]}
      stepped
      detailed
    />
  ),

  name: "Horizontal stepped",
};

export const Split: Story = {
  render: () => (
    <List
      items={[
        "Jointly shape the OpenStack architecture",
        "We help you plan your cloud hardware requirements",
        "We build OpenStack in your data center",
        "We operate the cloud to an SLA",
        "Transparent audit, logging, monitoring and management",
        "When your team is ready, we hand over the keys",
      ]}
      divided
      split
      ticked
    />
  ),

  name: "Split",
};
