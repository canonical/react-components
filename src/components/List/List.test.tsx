import { render, screen, within } from "@testing-library/react";
import React from "react";

import List from "./List";

describe("List ", () => {
  let items: string[];

  beforeEach(() => {
    items = ["test", "items", "here"];
  });

  it("renders", () => {
    render(<List items={items} />);
    expect(screen.getByRole("list")).toMatchSnapshot();
  });

  it("can define items as JSX", () => {
    render(<List items={[<button>test</button>, <span>items</span>]} />);
    expect(
      within(screen.getAllByRole("listitem")[0]).getByRole("button")
    ).toBeInTheDocument();
  });

  it("can define items as objects", () => {
    render(
      <List
        items={[
          {
            content: "test",
            className: "item-class",
          },
        ]}
      />
    );
    const item = within(screen.getByRole("listitem")).getByText("test");
    expect(item).toBeInTheDocument();
    expect(item).toHaveClass("item-class");
  });

  it("can be divided", () => {
    render(<List divided={true} items={items} />);
    expect(screen.getByRole("list")).toHaveClass("p-list--divided");
  });

  it("can be inline", () => {
    render(<List inline={true} items={items} />);
    expect(screen.getByRole("list")).toHaveClass("p-inline-list");
    expect(screen.getAllByRole("listitem")[0]).toHaveClass(
      "p-inline-list__item"
    );
  });

  it("can be inline middot", () => {
    render(<List middot={true} items={items} />);
    expect(screen.getByRole("list")).toHaveClass("p-inline-list--middot");
    expect(screen.getAllByRole("listitem")[0]).toHaveClass(
      "p-inline-list__item"
    );
  });

  it("can be inline stretch", () => {
    render(<List stretch={true} items={items} />);
    expect(screen.getByRole("list")).toHaveClass("p-inline-list--stretch");
    expect(screen.getByRole("list")).not.toHaveClass("p-list");
    expect(screen.getAllByRole("listitem")[0]).toHaveClass(
      "p-inline-list__item"
    );
  });

  it("can be split", () => {
    render(<List split={true} items={items} />);
    expect(screen.getByRole("list")).toHaveClass("is-split");
  });

  it("can be ticked", () => {
    render(<List ticked={true} items={items} />);
    expect(screen.getAllByRole("listitem")[0]).toHaveClass("is-ticked");
  });

  it("can be stepped", () => {
    render(
      <List
        stepped={true}
        items={[
          {
            title: "Test title",
            content: "test",
            className: "item-class",
            titleElement: "h1",
          },
        ]}
      />
    );
    expect(screen.getByRole("list")).toHaveClass("p-stepped-list");
    const item = screen.getAllByRole("listitem")[0];
    expect(item).toHaveClass("p-stepped-list__item");
    expect(within(item).getByText("Test title")).toHaveClass(
      "p-stepped-list__title"
    );
    expect(within(item).getByText("test")).toHaveClass(
      "p-stepped-list__content"
    );
  });

  it("can be detailed stepped", () => {
    render(
      <List
        detailed={true}
        stepped={true}
        items={[
          {
            title: "Test title",
            content: "test",
            className: "item-class",
            titleElement: "h1",
          },
        ]}
      />
    );
    expect(screen.getByRole("list")).toHaveClass("p-stepped-list--detailed");
    const item = screen.getAllByRole("listitem")[0];
    expect(item).toHaveClass("p-stepped-list__item");
    expect(within(item).getByText("Test title")).toHaveClass(
      "p-stepped-list__title"
    );
    expect(within(item).getByText("test")).toHaveClass(
      "p-stepped-list__content"
    );
  });

  it("can add additional classes", () => {
    render(<List className="extra-class" items={items} />);
    const list = screen.getByRole("list");
    expect(list).toHaveClass("p-list");
    expect(list).toHaveClass("extra-class");
  });
});
