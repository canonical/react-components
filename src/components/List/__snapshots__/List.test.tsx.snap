// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List  renders 1`] = `
<ul
  class="p-list"
>
  <li
    class="p-list__item"
  >
    test
  </li>
  <li
    class="p-list__item"
  >
    items
  </li>
  <li
    class="p-list__item"
  >
    here
  </li>
</ul>
`;
